messaging/messaging.go

```diff
@@ -0,0 +1,454 @@
+		s.Alert = a.Alert
+	} else {
+		s.Alert = a.AlertString
+	}
+	if a.ContentAvailable {
+		one := 1
+		s.ContentAvailable = &one
+	}
+	return json.Marshal(s)
+}
+
+// ApsAlert is the alert payload that can be included in an Aps.
+//
+// See https://developer.apple.com/library/content/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/PayloadKeyReference.html
+// for supported fields.
+type ApsAlert struct {
+	Title        string   `json:"title,omitempty"` // if specified, overrides the Title field of the Notification type
+	Body         string   `json:"body,omitempty"`  // if specified, overrides the Body field of the Notification type
+	LocKey       string   `json:"loc-key,omitempty"`
+	LocArgs      []string `json:"loc-args,omitempty"`
+	TitleLocKey  string   `json:"title-loc-key,omitempty"`
+	TitleLocArgs []string `json:"title-loc-args,omitempty"`
+	ActionLocKey string   `json:"action-loc-key,omitempty"`
+	LaunchImage  string   `json:"launch-image,omitempty"`
+}
+
+// ErrorInfo is a topic management error.
+type ErrorInfo struct {
+	Index  int
+	Reason string
+}
+
+// TopicManagementResponse is the result produced by topic management operations.
+//
+// TopicManagementResponse provides an overview of how many input tokens were successfully handled,
+// and how many failed. In case of failures, the Errors list provides specific details concerning
+// each error.
+type TopicManagementResponse struct {
+	SuccessCount int
+	FailureCount int
+	Errors       []*ErrorInfo
+}
+
+func newTopicManagementResponse(resp *iidResponse) *TopicManagementResponse {
+	tmr := &TopicManagementResponse{}
+	for idx, res := range resp.Results {
+		if len(res) == 0 {
+			tmr.SuccessCount++
+		} else {
+			tmr.FailureCount++
+			code, _ := res["error"].(string)
+			reason := iidErrorCodes[code]
+			if reason == "" {
+				reason = "unknown-error"
+			}
+			tmr.Errors = append(tmr.Errors, &ErrorInfo{
+				Index:  idx,
+				Reason: reason,
+			})
+		}
+	}
+	return tmr
+}
+
+// NewClient creates a new instance of the Firebase Cloud Messaging Client.
+//
+// This function can only be invoked from within the SDK. Client applications should access the
+// the messaging service through firebase.App.
+func NewClient(ctx context.Context, c *internal.MessagingConfig) (*Client, error) {
+	if c.ProjectID == "" {
+		return nil, errors.New("project id is required to access firebase cloud messaging client")
+	}
+
+	hc, _, err := transport.NewHTTPClient(ctx, c.Opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		fcmEndpoint: messagingEndpoint,
+		iidEndpoint: iidEndpoint,
+		client:      &internal.HTTPClient{Client: hc},
+		project:     c.ProjectID,
+		version:     "Go/Admin/" + c.Version,
+	}, nil
+}
+
+// Send sends a Message to Firebase Cloud Messaging.
+//
+// The Message must specify exactly one of Token, Topic and Condition fields. FCM will
+// customize the message for each target platform based on the parameters specified in the
```

Comment:
Super pedantic nit: https://stackoverflow.com/questions/156767/whats-the-difference-between-an-argument-and-a-parameter

Fine as-is.

Reply:
Changed to arguments